File: Android/tuilivekit/livekit/src/main/java/com/trtc/uikit/livekit/features/livelist/LiveListView.java
package com.trtc.uikit.livekit.features.livelist;

import android.content.Context;
import android.util.AttributeSet;
import android.widget.FrameLayout;

import androidx.annotation.NonNull;
import androidx.fragment.app.FragmentActivity;

import com.trtc.uikit.livekit.features.livelist.LiveListViewDefine.LiveListViewAdapter;
import com.trtc.uikit.livekit.features.livelist.LiveListViewDefine.Style;
import com.trtc.uikit.livekit.features.livelist.access.DoubleColumnListViewAdapter;
import com.trtc.uikit.livekit.features.livelist.access.SingleColumnListViewAdapter;
import com.trtc.uikit.livekit.features.livelist.access.TUILiveListDataSource;
import com.trtc.uikit.livekit.features.livelist.manager.LiveInfoListService;
import com.trtc.uikit.livekit.features.livelist.view.doublecolumn.DoubleColumnListView;
import com.trtc.uikit.livekit.features.livelist.view.singlecolumn.SingleColumnListView;

public class LiveListView extends FrameLayout {
    private LiveListViewDefine.Style mStyle;
    private SingleColumnListView     mSingleColumnListView;
    private DoubleColumnListView     mDoubleColumnListView;

    private LiveInfoListService                  mLiveInfoListService;
    private FragmentActivity                     mFragmentActivity;
    private LiveListViewAdapter                  mLiveListViewAdapter;
    private LiveListViewDefine.OnItemClickListener mOnItemClickListener;
    private boolean                              mIsInit;

    public LiveListView(@NonNull Context context) {
        this(context, null, 0);
    }

    public LiveListView(Context context, AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public LiveListView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
    }

    public void init(@NonNull FragmentActivity fragmentActivity, LiveListViewDefine.Style style) {
        init(fragmentActivity, style, null, null);
    }

    public void init(@NonNull FragmentActivity fragmentActivity, LiveListViewDefine.Style style,
                     LiveListViewAdapter adapter, LiveListViewDefine.LiveListDataSource dataSource) {
        mFragmentActivity = fragmentActivity;
        mStyle = style;
        mLiveListViewAdapter = adapter;
        if (dataSource == null) {
            dataSource = new TUILiveListDataSource();
        }
        mLiveInfoListService = new LiveInfoListService(dataSource);
        initLiveColumnListView(fragmentActivity, style, mLiveInfoListService, adapter);
        mIsInit = true;
    }

    public void updateColumnStyle(LiveListViewDefine.Style style) {
        mStyle = style;
        if (mIsInit) {
            initLiveColumnListView(mFragmentActivity, mStyle, mLiveInfoListService, mLiveListViewAdapter);
        }
    }

    public void setOnItemClickListener(LiveListViewDefine.OnItemClickListener listener) {
        mOnItemClickListener = listener;
        if (mStyle == Style.DOUBLE_COLUMN) {
            setDoubleColumnListViewClickLister(listener);
        } else {
            setSingleColumnListViewClickLister(listener);
        }
    }

    public void refreshData() {
        if (mStyle == Style.DOUBLE_COLUMN) {
            if (mDoubleColumnListView != null) {
                mDoubleColumnListView.refreshData();
            }
        } else {
            if (mSingleColumnListView != null) {
                mSingleColumnListView.refreshData();
            }
        }
    }

    private void initLiveColumnListView(@NonNull FragmentActivity fragmentActivity,
                                        @NonNull LiveListViewDefine.Style style,
                                        @NonNull LiveInfoListService liveInfoListService,
                                        LiveListViewAdapter adapter) {
        if (adapter == null) {
            if (style == Style.DOUBLE_COLUMN) {
                adapter = new DoubleColumnListViewAdapter(fragmentActivity);
            } else {
                adapter = new SingleColumnListViewAdapter(fragmentActivity);
            }
        }

        removeAllViews();
        if (style == Style.DOUBLE_COLUMN) {
            mDoubleColumnListView = new DoubleColumnListView(fragmentActivity);
            mDoubleColumnListView.init(fragmentActivity, adapter, liveInfoListService);
            addView(mDoubleColumnListView);
            setDoubleColumnListViewClickLister(mOnItemClickListener);
        } else {
            mSingleColumnListView = new SingleColumnListView(fragmentActivity);
            mSingleColumnListView.init(fragmentActivity, adapter, liveInfoListService);
            addView(mSingleColumnListView);
            setSingleColumnListViewClickLister(mOnItemClickListener);
        }
    }

    private void setDoubleColumnListViewClickLister(LiveListViewDefine.OnItemClickListener listener) {
        if (mDoubleColumnListView != null) {
            mDoubleColumnListView.setOnItemClickListener(listener);
        }
    }

    private void setSingleColumnListViewClickLister(LiveListViewDefine.OnItemClickListener listener) {
        if (mSingleColumnListView != null) {
            mSingleColumnListView.setOnItemClickListener(listener);
        }
    }
}

File: Android/tuilivekit/src/main/java/com/trtc/uikit/livekit/features/livelist/LiveListView.kt
package com.trtc.uikit.livekit.features.livelist

import android.content.Context
import android.util.AttributeSet
import android.widget.FrameLayout
import androidx.fragment.app.FragmentActivity
import com.tencent.qcloud.tuicore.TUICore
import com.trtc.uikit.livekit.features.livelist.access.DoubleColumnListViewAdapter
import com.trtc.uikit.livekit.features.livelist.access.SingleColumnListViewAdapter
import com.trtc.uikit.livekit.features.livelist.access.TUILiveListDataSource
import com.trtc.uikit.livekit.features.livelist.manager.LiveInfoListService
import com.trtc.uikit.livekit.features.livelist.view.doublecolumn.DoubleColumnListView
import com.trtc.uikit.livekit.features.livelist.view.singlecolumn.SingleColumnListView

class LiveListView @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = 0
) : FrameLayout(context, attrs, defStyleAttr) {

    private lateinit var style: Style
    private lateinit var singleColumnListView: SingleColumnListView
    private lateinit var doubleColumnListView: DoubleColumnListView

    private lateinit var liveInfoListService: LiveInfoListService
    private lateinit var fragmentActivity: FragmentActivity
    private var liveListViewAdapter: LiveListViewAdapter? = null
    private var onItemClickListener: OnItemClickListener? = null
    private var isInit = false

    fun init(
        fragmentActivity: FragmentActivity,
        style: Style,
        adapter: LiveListViewAdapter? = null,
        dataSource: LiveListDataSource? = null
    ) {
        this.fragmentActivity = fragmentActivity
        this.style = style
        liveListViewAdapter = adapter
        val liveDataSource = dataSource ?: TUILiveListDataSource()
        liveInfoListService = LiveInfoListService(liveDataSource)
        initLiveColumnListView(fragmentActivity, style, liveInfoListService)
        isInit = true
    }

    fun updateColumnStyle(style: Style) {
        if (isInit && style != this.style) {
            initLiveColumnListView(fragmentActivity, style, liveInfoListService)
        }
        this.style = style
    }

    fun setOnItemClickListener(listener: OnItemClickListener) {
        onItemClickListener = listener
        when (style) {
            Style.DOUBLE_COLUMN -> setDoubleColumnListViewClickLister(listener)
            else -> setSingleColumnListViewClickLister(listener)
        }
    }

    fun refreshData() {
        when (style) {
            Style.DOUBLE_COLUMN -> doubleColumnListView.refreshData()
            else -> singleColumnListView.refreshData()
        }
    }

    override fun onAttachedToWindow() {
        super.onAttachedToWindow()
        enableSwitchPlaybackQuality(true)
    }

    override fun onDetachedFromWindow() {
        super.onDetachedFromWindow()
        enableSwitchPlaybackQuality(false)
    }

    private fun initLiveColumnListView(
        fragmentActivity: FragmentActivity,
        style: Style,
        liveInfoListService: LiveInfoListService
    ) {
        val adapter = liveListViewAdapter ?: when (style) {
            Style.DOUBLE_COLUMN -> DoubleColumnListViewAdapter(fragmentActivity)
            else -> SingleColumnListViewAdapter(fragmentActivity)
        }

        removeAllViews()
        if (style == Style.DOUBLE_COLUMN) {
            doubleColumnListView = DoubleColumnListView(fragmentActivity).apply {
                init(fragmentActivity, adapter, liveInfoListService)
            }
            addView(doubleColumnListView)
            setDoubleColumnListViewClickLister(onItemClickListener)
        } else {
            singleColumnListView = SingleColumnListView(fragmentActivity).apply {
                init(fragmentActivity, adapter, liveInfoListService)
            }
            addView(singleColumnListView)
            setSingleColumnListViewClickLister(onItemClickListener)
        }
    }

    private fun setDoubleColumnListViewClickLister(listener: OnItemClickListener?) {
        listener?.let { doubleColumnListView.setOnItemClickListener(it) }
    }

    private fun setSingleColumnListViewClickLister(listener: OnItemClickListener?) {
        listener?.let { singleColumnListView.setOnItemClickListener(it) }
    }

    private fun enableSwitchPlaybackQuality(enable: Boolean) {
        val params: MutableMap<String?, Any?> = HashMap<String?, Any?>()
        params.put("enable", enable)
        TUICore.callService("AdvanceSettingManager", "enableSwitchPlaybackQuality", params)
    }
}
